### PR TITLE
Implement collectWithIndex on ListIterate and RandomAccessListIterate.

### PR DESCRIPTION
**Pull request by donraab**
Signed-off-by: Donald Raab <Donald.Raab@bnymellon.com>

### CODE REVIEW

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ListIterate.java [hunk 1]
```diff
@@ -357,6 +358,36 @@ private static boolean nonRandomAccessEquals(List<?> one, List<?> two)
     }
 
     /**
+     * @since 9.1
+     */
+    public static <T, A> MutableList<A> collectWithIndex(
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ListIterate.java]: In a utility like ListIterate is it better to provide only the target collection variant?

File: unit-tests/src/test/java/org/eclipse/collections/impl/utility/ListIterateTest.java [hunk 2]
```diff
@@ -192,6 +196,30 @@ private void assertCollect(List<Boolean> list)
         Verify.assertListsEqual(newCollection2, FastList.newListWith("true", "false", "null"));
     }
 
+    /**
+     * @since 9.1.
+     */
+    @Test
+    public void collectWithIndex()
+    {
+        MutableList<Boolean> list = Lists.fixedSize.of(true, false, null);
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/utility/ListIterateTest.java]: Can you please add 1 more test case which uses ArrayList<>()?

**Comment by donraab** [comment]
I disagree.  I did not know that OrderedIterate existed, but looking at it, it has one method that expects OrderedIterable.  I am limiting this pull request specifically to ListIterate and RandomAccessListIterate intentionally.  Anything else is out of scope for this work.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ListIterate.java (hunk 1 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/ListIterate.java]: Good question.  The non-target collection variant is more generally useful when used as utility, which is why I included it.

File: unit-tests/src/test/java/org/eclipse/collections/impl/utility/ListIterateTest.java (hunk 2 above)
Comment (donraab) [reply unit-tests/src/test/java/org/eclipse/collections/impl/utility/ListIterateTest.java]: Yes, I will add a test.

**Comment by nikhilnanivadekar** [comment]
ok fair enough. But if we add `collectWithIndex()` on an interface, it will go on `OrderedIterable`

**Comment by donraab** [comment]
Yes, I am planning on adding the method to the `OrderedIterable` interface.

**Comment by nikhilnanivadekar** [comment]
Raised issue #395

**Pull request merged**